app/scripts/lib/network-controller/messenger-action-handlers.ts

```diff
@@ -0,0 +1,152 @@
+}): void {
+  const isInfuraEndpointUrl = getIsInfuraEndpointUrl(
+    endpointUrl,
+    infuraProjectId,
+  );
+  const isQuicknodeEndpointUrl = getIsQuicknodeEndpointUrl(endpointUrl);
+  if (
+    (isInfuraEndpointUrl || isQuicknodeEndpointUrl) &&
+    !isConnectionError(error)
+  ) {
+    log.debug(
+      `Creating Segment event "${
+        MetaMetricsEventName.RpcServiceUnavailable
+      }" with chain_id_caip: "eip155:${chainId}", rpc_endpoint_url: ${onlyKeepHost(
+        endpointUrl,
+      )}`,
+    );
+    trackEvent({
+      category: MetaMetricsEventCategory.Network,
+      event: MetaMetricsEventName.RpcServiceUnavailable,
+      properties: {
+        chain_id_caip: `eip155:${hexToNumber(chainId)}`,
+        rpc_endpoint_url: onlyKeepHost(endpointUrl),
+      },
+    });
+  }
+}
+
+/**
+ * Handler for the `NetworkController:rpcEndpointDegraded` messenger action,
+ * which is called when an RPC endpoint is slow to return a successful response,
+ * or it cannot be reached or does not respond successfully after some number of
+ * retries.
+ *
+ * In this case, when we detect that Infura or Quicknode are degraded, we create
+ * an event in Segment so that we know to investigate further.
+ *
+ * @param args - The arguments.
+ * @param args.chainId - The chain ID that the endpoint represents.
+ * @param args.endpointUrL - The URL of the endpoint.
+ * @param args.infuraProjectId - Our Infura project ID.
+ * @param args.trackEvent - The function that will create the Segment event.
+ */
+export function onRpcEndpointDegraded({
+  chainId,
+  endpointUrl,
+  infuraProjectId,
+  trackEvent,
+}: {
+  chainId: Hex;
+  endpointUrl: string;
+  error: unknown;
+  infuraProjectId: string;
+  trackEvent: MetaMetricsController['trackEvent'];
+}): void {
+  const isInfuraEndpointUrl = getIsInfuraEndpointUrl(
+    endpointUrl,
+    infuraProjectId,
+  );
+  const isQuicknodeEndpointUrl = getIsQuicknodeEndpointUrl(endpointUrl);
+  if (isInfuraEndpointUrl || isQuicknodeEndpointUrl) {
+    log.debug(
+      `Creating Segment event "${
+        MetaMetricsEventName.RpcServiceDegraded
+      }" with chain_id_caip: "eip155:${chainId}", rpc_endpoint_url: ${onlyKeepHost(
+        endpointUrl,
+      )}`,
+    );
+    trackEvent({
+      category: MetaMetricsEventCategory.Network,
+      event: MetaMetricsEventName.RpcServiceDegraded,
+      properties: {
+        chain_id_caip: `eip155:${hexToNumber(chainId)}`,
+        rpc_endpoint_url: onlyKeepHost(endpointUrl),
+      },
+    });
+  }
+}
+
+/**
+ * Determines whether the given RPC endpoint URL matches an Infura URL that uses
+ * our API key.
+ *
+ * @param endpointUrl - The URL of the RPC endpoint.
+ * @param infuraProjectId - Our Infura project ID.
+ * @returns True if the URL is an Infura URL, false otherwise.
+ */
+function getIsInfuraEndpointUrl(
+  endpointUrl: string,
+  infuraProjectId: string,
+): boolean {
+  return new RegExp(
+    `^https://[^.]+.infura.io/v3/${escapeRegExp(infuraProjectId)}$`,
```

Comment:
this will match `https://notinfura.io/v3/<infuraProjectId>`

Reply:
True. I considered making this more strict, but it doesn't seem to me that we have a good way of checking what a valid Infura network is. The list of networks defined in `eth-json-rpc-infura` would be logical to use, but we only have a type there, not an enum/constant. `allowedInfuraHosts` in `shared/constants/network.ts` would also be logical but we only use that for a migration at the moment so who knows if it'll stick around.

I'm okay with this change for now. I suppose a user could add a new RPC endpoint that uses our API key and matches this pattern and yet uses a fake subdomain... but I don't know how common that would be. If we start seeing events in Segment for non-existent Infura networks then I think we should be able to filter them out pretty easily.

Reply:
I think you should just get the domain-check correct., subdomains don't matter so much.

Reply:
in case I wasn't clear: this allows any domain that ends with `infura.io`, not just any subdomain of `infura.io`

Reply:
Oh... sorry. Yes. Thanks for pointing that out. For some reason I thought I was already handling this. I'll fix that tomorrow.

Reply:
Fixed here: [60a3867](https://github.com/MetaMask/metamask-extension/pull/31089/commits/60a3867a9b44f86473fe68e45741191f50d6c5fb)